### PR TITLE
libc: add bsearch test

### PR DESCRIPTION
**Pull request by adamdebek**
JIRA: CI-239

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by damianloew — COMMENTED** [review]
What about edge cases like passing 0 as `nel`? I suppose that there are a little bit more things to test (make sure that every utility is tested and use as many edge cases as possible). Maybe it's not crucial to have many of tests for this function, but still please look through it and think about possible scenarios.

File: libc/bsearch.c [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	for (i = 0; i < TABSIZE; i++) {
+		node.string = names[i];
+		node.length = strlen(names[i]);
+		rval = (struct node *)bsearch((void *)&node, (void *)table,
+			(size_t)TABSIZE, sizeof(struct node), node_cmp);
+		TEST_ASSERT_NOT_NULL(rval);
+		TEST_ASSERT_EQUAL_MEMORY(&table[i], rval, sizeof(*rval));
+	}
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
```
Comment (damianloew) [inline libc/bsearch.c:141]: Why only name in each member is checked? The `length` element could potentially change.

Comment (damianloew) [reply libc/bsearch.c:141]: Above that, maybe it could be good to check whether members didn't change after each bsearch() call

File: libc/bsearch.c [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+	/* check for odd number of elements */
+	for (i = 0; i < TABSIZE - 1; i++) {
```
Comment (damianloew) [inline libc/bsearch.c:144]: If I understand correctly the order of elements could be changed, so it's not certain that the last member equals `names[TABSIZE]`

File: libc/bsearch.c [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+}
+
+
+TEST(bsearch, find_node_not_in)
+{
+	struct node *rval, node;
+	char name[8];
+
+	TEST_ASSERT_EQUAL_MEMORY(name, strcpy(name, "phoenix"), sizeof(name));
```
Comment (damianloew) [inline libc/bsearch.c:162]: We could just pass "notExistentMember" as key.

File: libc/bsearch.c [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	/* check for odd number of elements */
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		(size_t)TABSIZE - 1, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+}
+
+
+TEST(bsearch, not_sorted)
```
Comment (damianloew) [inline libc/bsearch.c:181]: We could also look for all the elements there.

File: libc/bsearch.c [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+
+TEST_GROUP_RUNNER(bsearch)
+{
+	RUN_TEST_CASE(bsearch, zero_el);
+	prepare(names);
+	RUN_TEST_CASE(bsearch, find_node_in);
+	RUN_TEST_CASE(bsearch, find_node_not_in);
+	prepare(names_nsort);
+	RUN_TEST_CASE(bsearch, not_sorted);
```
Comment (damianloew) [inline libc/bsearch.c:212]: It's not checked:
Screenshot from 2023-06-07 18-29-45

Two or more members comapre equal - please add such case.

**Review by damianloew — CHANGES_REQUESTED** [review]
We could certainly add cases, where key element meets the `compar` requirements partially, so for example when we decided to use simple strcmp() we would want to have such key: "bc\0d", when the members on `base` address are "abc", "def", and so on. We may want to make sure that members are correctly splitted and passed to the `compar` function.

File: libc/bsearch.c (hunk 4 above)
Comment (adamdebek) [reply libc/bsearch.c:181]: To do this i would need to prepare test_table per every elements which is unnecessary effort in my opinion.

File: libc/stdlib_bsearch.c [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+{
+	int i;
+
+	for (i = 0; i < TABSIZE; i++) {
+		test_table[i].string = (char *)malloc(strlen(names_arg[i]) + 1);
+
+		TEST_ASSERT_NOT_NULL(test_table[i].string);
+
+		strcpy(test_table[i].string, names_arg[i]);
+		test_table[i].length = strlen(names_arg[i]);
```
Comment (damianloew) [inline libc/stdlib_bsearch.c:85]: Please call strlen one time in the iteration.

File: libc/stdlib_bsearch.c [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+/* compare function used by bsearch */
+static int test_nodeCmp(const void *node1, const void *node2)
+{
+	return strcmp(((struct node *)node1)->string, ((struct node *)node2)->string);
+}
+
+
+/* compare function used by bsearch in special_node testcase */
+static int test_nodeCmp2(const void *node1, const void *node2)
```
Comment (damianloew) [inline libc/stdlib_bsearch.c:131]: Please use more distinctive names.

```suggestion
static int test_nodeCmpStr(const void *node1, const void *node2)
{
	return strcmp(((struct node *)node1)->string, ((struct node *)node2)->string);
}


/* compare function used by bsearch in special_node testcase */
static int test_nodeCmpLen(const void *node1, const void *node2)
```

File: libc/stdlib_bsearch.c [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+	char name[] = "adam";
+	node2.string = name;
+	node2.length = strlen(name);
+
+	/* Check for even number of elements */
+	test_prepare(names);
+
+	rval = (struct node *)bsearch((void *)&node2, (void *)test_table,
+		(size_t)TABSIZE, sizeof(struct node), test_nodeCmp2);
+	TEST_ASSERT_NULL(rval);
```
Comment (damianloew) [inline libc/stdlib_bsearch.c:305]: Please explain this case a little bit in a comment, the best would be o provide and example, so sth like: 

```C
/* the table content is {str1len1str2len2str3len3}, so we assert that elements are splitted properly and len1str2 won't be found */
```

**Review by damianloew — CHANGES_REQUESTED** [review]

**Review by damianloew — APPROVED** [review]
